R/PipeOpImputeMode.R

```diff
@@ -0,0 +1,88 @@
+#' @title PipeOpImputeMode
+#'
+#' @usage NULL
+#' @name mlr_pipeops_imputemode
+#' @format [`R6Class`] object inheriting from [`PipeOpImpute`]/[`PipeOp`].
+#'
+#' @description
+#' Impute numerical features by their mode.
+#'
+#' @section Construction:
+#' ```
+#' PipeOpImputeMode$new(id = "imputemode", param_vals = list())
+#' ```
+#'
+#' * `id` :: `character(1)`\cr
+#'   Identifier of resulting object, default `"imputemode"`.
+#' * `param_vals` :: named `list`\cr
+#'   List of hyperparameter settings, overwriting the hyperparameter settings that would otherwise be set during construction. Default `list()`.
+#'
+#' @section Input and Output Channels:
+#' Input and output channels are inherited from [`PipeOpImpute`].
+#'
+#' The output is the input [`Task`][mlr3::Task] with all affected numeric features missing values imputed by (column-wise) mode.
+#'
+#' @section State:
+#' The `$state` is a named `list` with the `$state` elements inherited from [`PipeOpImpute`].
+#'
+#' The `$state$model` is a named `list` of `numeric(1)` indicating the mode of the respective feature.
+#'
+#' @section Parameters:
+#' The parameters are the parameters inherited from [`PipeOpImpute`], as well as the following
+#' parameters based on [`compute_mode()`][mlr3misc::compute_mode]:
+#' * `ties_method` :: `character(1)`\cr
+#'   Ties handling type. One of \dQuote{random} (default), \dQuote{first} or \dQuote{last}.
+#'
+#' @section Internals:
+#' Uses the [`mlr3misc::compute_mode()`] function. Features that are entirely `NA` are imputed as `0`.
+#'
+#' @section Methods:
+#' Only methods inherited from [`PipeOpImpute`]/[`PipeOp`].
+#'
+#' @examples
+#' library("mlr3")
+#'
+#' task = tsk("pima")
+#' task$missings()
+#'
+#' po = po("imputemode")
+#' new_task = po$train(list(task = task))[[1]]
+#' new_task$missings()
+#'
+#' po$state$model
+#' @family PipeOps
+#' @family Imputation PipeOps
+#' @include PipeOpImpute.R
+#' @export
+PipeOpImputeMode = R6Class("PipeOpImputeMode",
+  inherit = PipeOpImpute,
+  public = list(
+    initialize = function(id = "imputemode", param_vals = list()) {
+      ps = ParamSet$new(params = list(
+        ParamFct$new("ties_method", levels = c("first", "last", "random"), default = "random", tags = c("train", "predict"))
+      ))
+      ps$values = list(ties_method = "random")
+      super$initialize(id, param_set = ps, param_vals = param_vals, packages = "mlr3misc")
+    },
+
+    select_cols = function(task) task$feature_types[get("type") %in% c("numeric", "integer"), get("id")],
+
+    train_imputer = function(feature, type, context) {
+      mod = mlr3misc::compute_mode(feature, ties_method = self$param_set$values$ties_method)
+      if (is.na(mod)) {
+        mod = 0
```

Comment:
1. Cannot be NA unless `na_rm = TRUE` is passed to `compute_mode()`
2. The mode should be one of the observed values, imputing 0 is not reasonable.
3. It is not clear what value to impute if all values of an integer are NA. What are we doing here while imputing with the mean? -- for factors, we could just sample a random factor level.

Reply:
1. fixed, thanks.

2. and 3. `PipeOpMean` and `PipeOpMedian` both always impute `0` for numerics or integers if all are `NA`. For `factor`s (also `ordered`s) `PipeOpImputeMode` now samples a level randomly. (Does it make sense to modify the behavior here to also match the `ties_method` argument? I.e. if all values are `NA`, then either a random level is sampled - `random`, or the `first` or `last` level is choosen). Logicals are now also supported (if all values are `NA` `TRUE`/`FALSE` is being sampled).

Reply:
> Logicals are now also supported (if all values are `NA` `TRUE`/`FALSE` is being sampled).

Great!


>   `PipeOpMean` and `PipeOpMedian` both always impute `0` for numerics or integers if all are `NA`

Okay ... I find it odd, but at least it is properly documented. Then we should do the same thing here.

> Does it make sense to modify the behavior here to also match the ties_method argument? I.e. if all values are NA, then either a random level is sampled - random, or the first or last level is choosen

Yes, makes sense.

Reply:
> Okay ... I find it odd, but at least it is properly documented. Then we should do the same thing here.

This also the default for `PipeOpImputeMode` now and it is also documented.

> Yes, makes sense.

Okay, fixed this. Maybe @mb706 wants to leave a final comment regarding the tests!